derp/derp_server.go

```diff
@@ -1931,6 +1958,43 @@ func (s *Server) ServeDebugTraffic(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+// Print a list of data flows that have gone through this server
+// between connected clients. Includes clients which are currently
+// communicating directly but still using this DERP server as their
+// home.
+//
+// TODO: What else do we want to know? Time connection was
+// established? Time since last packet? Average bandwidth?
+//
+// TODO: Convert into Prometheus histograms
+func (s *Server) ServeDebugFlows(w http.ResponseWriter, r *http.Request) {
+	// Lock, then copy out the current map, then unlock
+	fmt.Fprintf(w, "src dst bytes packets\n")
+	s.mu.Lock()
+	// Copy the whole map into our sortable data structure so we
+	// can drop the server lock quickly
+	type flowLine struct {
+		src, dst key.NodePublic
+		flw      flowData
+	}
+	flows := make([]flowLine, 0)
+	for src := range s.sentTo {
+		for dst, flw := range s.sentTo[src] {
+			flows = append(flows, flowLine{src, dst, flw})
+		}
+	}
+	s.mu.Unlock()
+
+	// Sort by largest flow
+	sort.Slice(flows, func(i, j int) bool { return flows[i].flw.bytes < flows[j].flw.bytes })
+
+	for _, f := range flows {
+		fmt.Fprintf(w, "%v %v %v %v\n", f.src.ShortString(), f.dst.ShortString(), f.flw.bytes, f.flw.pkts)
+	}
+
+	w.(http.Flusher).Flush()
```

Comment:
not needed. It happens on exit. Plus you're assuming the ResponseWriter is a Flusher here.